src/servers/Server.cc

```diff
@@ -125,9 +134,21 @@ Server::doClientRead(const CommIoCbParams &io)
      * Plus, it breaks our lame *HalfClosed() detection
      */
 
+    // mayBufferMoreRequestBytes() was true during readSomeData(), but variables
+    // like Config.maxRequestBufferSize may have changed since that check
+    if (!mayBufferMoreRequestBytes()) {
+        // XXX: If we avoid Comm::ReadNow(), we should not Comm::Read() again
+        // when the wait is over; resume these doClientRead() checks instead.
+        return; // wait for noteMoreBodySpaceAvailable() or a similar inBuf draining event
+    }
     maybeMakeSpaceAvailable();
```

Comment:
The new condition below will read maximum rd.size bytes, and the question is whether maybeMakeSpaceAvailable() call may (vainly) reserve bytes that are not used due to the rd.size limitation.

The method sets up three parameters: requirements.minSpace=1024,  requirements.idealSpace=CLIENT_REQ_BUF_SZ=4096 and   requirements.maxCapacity=Config.maxRequestBufferSize. If inBuf.length() is much less than Config.maxRequestBufferSize, it reserves the max of the first two values (i.e., the buffer grows by 4096 each time it becomes full), and rd.size does not limit anything. When it comes near to Config.maxRequestBufferSize, inBuf capacity grows by Config.maxRequestBufferSize-inBuf.length() (notwithstanding extra byte(s) adding by cow()) - i.e., the value limited by rd.size. So in both cases we do not (vainly) reserve memory that is not used due to the rd.size condition - this condition just helps to avoid using the extra bytes that were implicitly added to the capacity by cow().